haystack/nodes/other/shaper.py

```diff
@@ -0,0 +1,364 @@
+        :type node: dict[str, Any]
+        :raises ValueError: If function is not registered in the function registry.
+        """
+        function_name = node["func"]
+        func = self.function_registry.get(function_name)
+        if not func:
+            raise ValueError(f"{function_name} not supported by Shaper. Check the function name and try again.")
+        return func
+
+    def traverse(self, parent, node, invocation_context, node_visitor):
+        """
+        Traverses the input tree in post-order and invokes the given node visitor function on each node.
+
+        See Shaper invoke function for more details.
+
+        :param parent: The name of the parent node.
+        :type parent: str
+        :param node: A dictionary or list representing the node.
+        :type node: Union[dict, list]
+        :param invocation_context: A dictionary of variables available in the invocation context.
+        :type invocation_context: dict[str, Any]
+        :param node_visitor: A function to invoke on each node.
+        :type node_visitor: Callable[[str, Union[dict, list], dict[str, Any]], None]
+        """
+        # traverse the inputs tree in post-order
+        if isinstance(node, dict):
+            for key, value in node.items():
+                self.traverse(key, value, invocation_context, node_visitor)
+        elif isinstance(node, list):
+            for item in node:
+                self.traverse(node, item, invocation_context, node_visitor)
+
+        if isinstance(node, Dict):
+            node_visitor(parent, node, invocation_context)
+
+    def run(
+        self,
+        query: Optional[str] = None,
+        file_paths: Optional[List[str]] = None,
+        labels: Optional[MultiLabel] = None,
+        documents: Optional[List[Document]] = None,
+        meta: Optional[dict] = None,
+    ) -> Tuple[Dict, str]:
+
+        invocation_context: Dict[str, Any] = {
+            "invocation_stack": deque(),
+            **{
+                k: v
+                for k, v in {"query": query, "file_paths": file_paths, "labels": labels, "documents": documents}.items()
+                if v is not None
+            },
+        }
+
+        # check if the inputs are valid (i.e. in invocation context)
+        invalid_vars = [key for key in self.inputs.keys() if key not in invocation_context]
+        if invalid_vars:
+            raise ValueError(
+                f"The following variables were not found in the invocation context {invalid_vars}."
+                f"Please check the inputs definition and try again."
+            )
+
+        # traverse the inputs YAML subtree of the Shaper component declaration, invoke the specified
+        # functions, and store the results in the invocation context
+        self.traverse(None, self.inputs, invocation_context, self.invoke)
+        invocation_context.pop("invocation_stack")
+
+        meta = meta or {}
+        meta["invocation_context"] = invocation_context
```

Comment:
Isn't this meta in the pipeline because it gets added to documents in nodes like the `PDFConverter`? Is it safe to just add things to this meta? Will it be added to document meta potentially? I think that would be bad because it would leak internal implementation details of the Shaper node.

Reply:
Or does the `PromptNode` also depend on the presence of this invocation context in meta?

Reply:
Yeah, it does; PromptNode looks for invocation context variables under meta["invocation_context"] + run parameters. This is a workaround for the current pipeline run method that hardcodes parameters (query, file_paths, labels, documents, meta etc) and doesn't allow kwargs.

Reply:
And what about the issue with changing the meta? Is it sure that this will not break other components which depend on meta later in the pipeline and that it won't get returned at the end of the pipeline?

Reply:
No, I did not mean that other components define something under `meta[invocation_context]`, I rather meant that some other component puts that meta onto a document and then the invocation_context ends up in the document store. But maybe we have to live with that.

Reply:
And a question regarding the `run`-method of the PromptNode. I have used other arguments and kwargs in custom nodes. What prevents you from specifying in the `run`-method of the `PromptNode` that it should accept kwargs? I know you would change the signature compared to the base class but this seemed to work for me.

Reply:
> No, I did not mean that other components define something under `meta[invocation_context]`, I rather meant that some other component puts that meta onto a document and then the invocation_context ends up in the document store. But maybe we have to live with that.

Ahaa, we can put a flag on PromptNode to clean up meta["invocation_context"] and we could flip this switch on for some nodes? I'll talk to the core about this. TY

Reply:
> And a question regarding the `run`-method of the PromptNode. I have used other arguments and kwargs in custom nodes. What prevents you from specifying in the `run`-method of the `PromptNode` that it should accept kwargs? I know you would change the signature compared to the base class but this seemed to work for me.

I can specify them with no problem, the issue is that the Pipeline you load from YAML can't pass down the pipeline any other parameter. I'll also talk to core about this to see if we can unrestrict this. This would make the need for invocation context in meta obsolete. I'll get back to you.

Reply:
Ok @mathislucka, here is an update. After consultation with @ZanSara , we can drop the use of meta for invocation_context k,v pairs. We can use a declared `invocation_context` run method parameter in both PromptNode and Shaper. I've tested multiple PN nodes with Shaper(s) in between, and everything works. Thanks for pushing for clarifications on this front.

Reply:
Having said that, we can't (without some kludge or pipeline rework) allow arbitrary kwargs pipeline injection in the pipeline run method. So you are not going to be able to do something like this: 
```
pipeline = Pipeline.load_from_yaml(path=tmp_path / "tmp_config.yml")
result = pipeline.run(query="Some query here", documents=[Document("Berlin is an amazing city."), questions_to_ask=["Here q1", "Here is q2"]])
```

Please think about this - if it will be a significant limitation for dC. If not, it can wait for pipeline rework which should come around March or so.

Reply:
Another obvious consequence of this small change - no other components in the pipeline between Shapers and PromptNodes will be able to access variables in the invocation context. But I think we are ok with that, it's even advantageous.

Reply:
Ok, so I would add the `invocation_context` to the `run`-method for `PromptNode` and `PromptTemplate`, but as said in my other comments, I would actually return shaper output that overrides standard parameters like `query`, `documents`, `meta`, `labels` or `files` directly and not write it to `invocation_context`. Otherwise, this will be less useful for use cases like query expansion.

Reply:
Of course, we need to think about making sure that we retain the type of these parameters. `query` should remain a string, `documents` should remain a list of documents, `meta` should remain a dict and so on